core/blockchain.go

```diff
@@ -2628,50 +2636,72 @@ func (bc *BlockChain) UpdateValidatorVotingPower(state *shard.State) error {
+		validator staking.ValidatorStats
+	}
 
-	for _, committee := range state.Shards {
-		for _, slot := range committee.Slots {
-			if slot.TotalStake != nil {
-				if _, ok := addrToEffectiveStakes[slot.EcdsaAddress]; !ok {
-					addrToEffectiveStakes[slot.EcdsaAddress] = map[uint32]numeric.Dec{}
-				}
-				if _, ok := addrToEffectiveStakes[slot.EcdsaAddress][committee.ShardID]; !ok {
-					addrToEffectiveStakes[slot.EcdsaAddress][committee.ShardID] = numeric.NewDec(0)
-				}
-				addrToEffectiveStakes[slot.EcdsaAddress][committee.ShardID] = addrToEffectiveStakes[slot.EcdsaAddress][committee.ShardID].Add(*slot.TotalStake)
+	stats := []withAddr{}
+	rosters := make([]*votepower.Roster, len(state.Shards))
 
-				if _, ok := totalEffectiveStake[committee.ShardID]; !ok {
-					totalEffectiveStake[committee.ShardID] = numeric.NewDec(0)
-				}
-				totalEffectiveStake[committee.ShardID] = totalEffectiveStake[committee.ShardID].Add(*slot.TotalStake)
-			}
+	for i := range state.Shards {
+		roster, err := votepower.Compute(state.Shards[i].Slots)
+		if err != nil {
+			return err
 		}
+		rosters[i] = roster
 	}
 
-	batch := bc.db.NewBatch()
-	for addr, votingPowers := range addrToEffectiveStakes {
-		addrTotalVotingPower := numeric.NewDec(0) // Total voting power is the average voting power across all shards
-		addrTotalEffectiveStake := numeric.NewDec(0)
-		for shardID, eStake := range votingPowers {
-			addrTotalVotingPower = addrTotalVotingPower.Add(eStake.Quo(totalEffectiveStake[shardID]))
-			addrTotalEffectiveStake = addrTotalEffectiveStake.Add(eStake)
-		}
-		// Retrieve the stats and update
-		stats, err := rawdb.ReadValidatorStats(bc.db, addr)
-		if stats == nil {
-			stats = &staking.ValidatorStats{big.NewInt(0), big.NewInt(0), big.NewInt(0), numeric.NewDec(0), numeric.NewDec(0)}
+	for i := range state.Shards {
+		shardID := state.Shards[i].ShardID
+		for _, slot := range state.Shards[i].Slots {
+			if slot.TotalStake != nil {
+				stat := withAddr{slot.EcdsaAddress, staking.ValidatorStats{}}
+				stat.validator.AvgVotingPower = numeric.ZeroDec()
+				stat.validator.TotalEffectiveStake = numeric.ZeroDec()
+				roster := rosters[i]
+				stat.validator.AvgVotingPower = stat.validator.AvgVotingPower.Add(
+					roster.Voters[slot.BlsPublicKey].EffectivePercent,
+				)
+				stat.validator.TotalEffectiveStake = stat.validator.TotalEffectiveStake.Add(
+					roster.Voters[slot.BlsPublicKey].RawStake,
+				)
+				stat.validator.VotingPowerPerShard = append(
+					stat.validator.VotingPowerPerShard, staking.VotePerShard{
+						shardID, roster.Voters[slot.BlsPublicKey].EffectivePercent,
+					},
+				)
+				stat.validator.BLSKeyPerShard = append(
+					stat.validator.BLSKeyPerShard, staking.KeyPerShard{
+						shardID, roster.Voters[slot.BlsPublicKey].Identity.Hex(),
+					},
+				)
+				stat.validator.AvgVotingPower = stat.validator.AvgVotingPower.Quo(
+					numeric.NewDec(int64(len(rosters))),
+				)
+				stats = append(stats, stat)
```

Comment:
This logic doesn't seems right. The stats should be by validator. but here you have a stats for each bls key. Also the voting power should be total voting power of the validator in each shard divided by 4. But since you have separated voting powers for each key, probably better just remove the avgVotingPower and let frontend compute it.

Reply:
I thought about that too, but wasn't sure what the right answer would be (about stats by validator vs stats for each bls key). So would it be then a group by on the earning one address, right?